File: app/src/main/java/com/ruleengine/index/AhoCorasick.java
package com.ruleengine.index;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.function.Consumer;

/**
 * Aho-Corasick automaton for efficient multi-pattern substring matching.
 *
 * <p>Uses a DFA with array-indexed transitions for ASCII characters
 * (direct {@code int[128]} lookup per state) and a {@link HashMap} fallback
 * for non-ASCII. After {@link #build()}, the goto function is fully completed
 * so search requires no failure-link chasing — just a single array lookup
 * per character.
 *
 * <p>Usage:
 * <ol>
 *   <li>Insert patterns with {@link #insert(String, Object)}</li>
 *   <li>Call {@link #build()} to construct the automaton</li>
 *   <li>Call {@link #search(String, Consumer)} or {@link #search(String)}</li>
 * </ol>
 *
 * @param <V> the type of values associated with each pattern
 */
public final class AhoCorasick<V> {

    private static final int ASCII_SIZE = 128;
    private static final int NO_STATE = -1;
    private static final int INITIAL_CAPACITY = 64;

    private final List<V> emptyPatternValues = new ArrayList<>();

    // --- Two-phase data structures ---
    // During insertion, Lists allow dynamic growth as states are allocated.
    // After build(), these are flattened into fixed-size arrays for cache-friendly
    // search-phase access (contiguous memory, no List indirection). The build-phase
    // Lists are then nulled to free memory.
    //
    // Each state has two tiers of transitions (same strategy as Trie.Node):
    //   - ASCII (0–127): direct int[] array per state for O(1) lookup
    //   - Non-ASCII:      HashMap<Character, Integer> fallback (often null)

    // Build-phase structures (nulled after build)
    private List<int[]> gotoRows = new ArrayList<>(INITIAL_CAPACITY);
    @SuppressWarnings("unchecked")
    private List<Map<Character, Integer>> extendedRows = new ArrayList<>(INITIAL_CAPACITY);
    private List<List<V>> buildOutput = new ArrayList<>(INITIAL_CAPACITY);
    private int stateCount = 0;

    private static final Object[] EMPTY_OUTPUT = new Object[0];

    // Search-phase arrays (populated by build, one entry per state)
    private int[][] gotoTable;
    @SuppressWarnings("unchecked")
    private Map<Character, Integer>[] extendedGoto;
    private Object[][] output;
    private boolean built = false;

    /** Creates a new empty automaton with the root state allocated. */
    public AhoCorasick() {
        allocateState(); // state 0 = root
    }

    /**
     * Inserts a pattern with an associated value into the automaton.
     *
     * @param pattern the pattern string
     * @param value   the value to return when this pattern is found
     * @throws IllegalStateException if called after {@link #build()}
     */
    public void insert(String pattern, V value) {
        if (built) {
            throw new IllegalStateException("Cannot insert after build()");
        }
        if (pattern.isEmpty()) {
            emptyPatternValues.add(value);
            return;
        }
        int state = 0;
        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);
            int next = getGoto(state, c);
            if (next == NO_STATE) {
                next = allocateState();
                setGoto(state, c, next);
            }
            state = next;
        }
        buildOutput.get(state).add(value);
    }

    /**
     * Constructs the automaton by computing failure links and completing
     * the DFA transition table.
     *
     * <p>After this call, no further inserts are allowed and the automaton
     * is ready for searching. The build proceeds in four phases:
     * <ol>
     *   <li>Initialize depth-1 states (direct children of root)</li>
     *   <li>Compute failure links for all deeper states via BFS</li>
     *   <li>Complete the DFA so every state has a transition for every character</li>
     *   <li>Flatten build-phase Lists into fixed-size arrays for search</li>
     * </ol>
     */
    public void build() {
        int[] failure = new int[stateCount];
        Queue<Integer> queue = new ArrayDeque<>();

        initDepthOneStates(failure, queue);
        computeFailureLinks(failure, queue);
        completeDfa(failure, queue);
        flattenToArrays();
    }

    /**
     * Sets failure links for root's direct children to 0 (root) and fills
     * missing root transitions with self-loops back to root.
     */
    private void initDepthOneStates(int[] failure, Queue<Integer> queue) {
        int[] rootRow = gotoRows.get(0);
        for (int c = 0; c < ASCII_SIZE; c++) {
            int child = rootRow[c];
            if (child == NO_STATE) {
                rootRow[c] = 0; // missing root transitions loop to root
            } else {
                failure[child] = 0;
                queue.add(child);
            }
        }
        Map<Character, Integer> rootExt = extendedRows.get(0);
        if (rootExt != null) {
            for (Map.Entry<Character, Integer> entry : rootExt.entrySet()) {
                failure[entry.getValue()] = 0;
                queue.add(entry.getValue());
            }
        }
    }

    /**
     * BFS from depth-1 states to compute failure links and merge output lists.
     * Each state's failure link points to the longest proper suffix of the
     * path leading to that state which is also a prefix of some pattern.
     */
    private void computeFailureLinks(int[] failure, Queue<Integer> queue) {
        while (!queue.isEmpty()) {
            int current = queue.poll();
            int[] row = gotoRows.get(current);

            for (int c = 0; c < ASCII_SIZE; c++) {
                int child = row[c];
                if (child != NO_STATE) {
                    failure[child] = followFailure(failure, current, (char) c);
                    mergeOutput(child, failure[child]);
                    queue.add(child);
                }
            }

            Map<Character, Integer> ext = extendedRows.get(current);
            if (ext != null) {
                for (Map.Entry<Character, Integer> entry : ext.entrySet()) {
                    int child = entry.getValue();
                    failure[child] = followFailure(failure, current, entry.getKey());
                    mergeOutput(child, failure[child]);
                    queue.add(child);
                }
            }
        }
    }

    /**
     * Completes the DFA so every state has an explicit transition for every
     * character. Missing transitions are filled from the failure state's
     * (already-completed) row. BFS order guarantees that failure states are
     * processed before their dependents.
     */
    private void completeDfa(int[] failure, Queue<Integer> queue) {
        // Seed BFS with root's direct children (skip root itself)
        int[] rootRow = gotoRows.get(0);
        for (int c = 0; c < ASCII_SIZE; c++) {
            int child = rootRow[c];
            if (child != 0) {
                queue.add(child);
            }
        }
        Map<Character, Integer> rootExt = extendedRows.get(0);
        if (rootExt != null) {
            queue.addAll(rootExt.values());
        }

        while (!queue.isEmpty()) {
            int current = queue.poll();
            int[] row = gotoRows.get(current);
            int[] failRow = gotoRows.get(failure[current]);

            for (int c = 0; c < ASCII_SIZE; c++) {
                if (row[c] == NO_STATE) {
                    row[c] = failRow[c]; // inherit from failure state
                } else {
                    queue.add(row[c]);
                }
            }

            inheritExtendedTransitions(current, failure[current]);
            enqueueExtendedChildren(current, queue);
        }
    }

    /** Inherits non-ASCII transitions from the failure state for any chars not already present. */
    private void inheritExtendedTransitions(int state, int failState) {
        Map<Character, Integer> failExt = extendedRows.get(failState);
        if (failExt == null) {
            return;
        }
        Map<Character, Integer> ext = extendedRows.get(state);
        if (ext == null) {
            ext = new HashMap<>(4);
            extendedRows.set(state, ext);
        }
        for (Map.Entry<Character, Integer> entry : failExt.entrySet()) {
            ext.putIfAbsent(entry.getKey(), entry.getValue());
        }
    }

    /** Adds non-root non-ASCII child states to the BFS queue. */
    private void enqueueExtendedChildren(int state, Queue<Integer> queue) {
        Map<Character, Integer> ext = extendedRows.get(state);
        if (ext != null) {
            for (int child : ext.values()) {
                if (child != 0) {
                    queue.add(child);
                }
            }
        }
    }

    /** Flattens build-phase Lists into arrays and releases build-phase memory. */
    @SuppressWarnings("unchecked")
    private void flattenToArrays() {
        gotoTable = gotoRows.toArray(new int[0][]);
        extendedGoto = extendedRows.toArray(new Map[0]);

        output = new Object[stateCount][];
        for (int i = 0; i < stateCount; i++) {
            List<V> list = buildOutput.get(i);
            output[i] = list.isEmpty() ? EMPTY_OUTPUT : list.toArray();
        }

        gotoRows = null;
        extendedRows = null;
        buildOutput = null;
        built = true;
    }

    /**
     * Searches the text and invokes the consumer for each matching value.
     *
     * @param text     the text to search
     * @param consumer called for each matching value
     * @throws IllegalStateException if {@link #build()} has not been called
     */
    @SuppressWarnings("unchecked")
    public void search(String text, Consumer<V> consumer) {
        if (!built) {
            throw new IllegalStateException("Must call build() before search()");
        }
        for (V v : emptyPatternValues) {
            consumer.accept(v);
        }
        int state = 0;
        for (int i = 0; i < text.length(); i++) {
            state = nextState(state, text.charAt(i));
            Object[] matches = output[state];
            for (int j = 0; j < matches.length; j++) {
                consumer.accept((V) matches[j]);
            }
        }
    }

    /**
     * Searches the text for all inserted patterns and returns their associated values.
     *
     * @param text the text to search
     * @return list of values for all patterns found in the text
     * @throws IllegalStateException if {@link #build()} has not been called
     */
    @SuppressWarnings("unchecked")
    public List<V> search(String text) {
        if (!built) {
            throw new IllegalStateException("Must call build() before search()");
        }
        List<V> result = new ArrayList<>();
        for (V v : emptyPatternValues) {
            result.add(v);
        }
        int state = 0;
        for (int i = 0; i < text.length(); i++) {
            state = nextState(state, text.charAt(i));
            Object[] matches = output[state];
            for (int j = 0; j < matches.length; j++) {
                result.add((V) matches[j]);
            }
        }
        return result;
    }

    private int allocateState() {
        int id = stateCount++;
        int[] row = new int[ASCII_SIZE];
        Arrays.fill(row, NO_STATE);
        gotoRows.add(row);
        extendedRows.add(null);
        buildOutput.add(new ArrayList<>());
        return id;
    }

    private int getGoto(int state, char c) {
        if (c < ASCII_SIZE) {
            return gotoRows.get(state)[c];
        }
        Map<Character, Integer> ext = extendedRows.get(state);
        if (ext == null) {
            return NO_STATE;
        }
        return ext.getOrDefault(c, NO_STATE);
    }

    private void setGoto(int state, char c, int target) {
        if (c < ASCII_SIZE) {
            gotoRows.get(state)[c] = target;
        } else {
            Map<Character, Integer> ext = extendedRows.get(state);
            if (ext == null) {
                ext = new HashMap<>(4);
                extendedRows.set(state, ext);
            }
            ext.put(c, target);
        }
    }

    /**
     * Walks the failure chain from {@code parent} to find the deepest state
     * that has a goto transition for {@code c}, then returns that target.
     */
    private int followFailure(int[] failure, int parent, char c) {
        int state = failure[parent];
        while (state != 0 && getGoto(state, c) == NO_STATE) {
            state = failure[state];
        }
        int target = getGoto(state, c);
        return target != NO_STATE ? target : 0;
    }

    private void mergeOutput(int state, int failState) {
        List<V> failOut = buildOutput.get(failState);
        if (failOut != null && !failOut.isEmpty()) {
            buildOutput.get(state).addAll(failOut);
        }
    }

    private int nextState(int state, char c) {
        if (c < ASCII_SIZE) {
            return gotoTable[state][c];
        }
        Map<Character, Integer> ext = extendedGoto[state];
        if (ext != null) {
            Integer next = ext.get(c);
            if (next != null) {
                return next;
            }
        }
        return 0; // unknown non-ASCII char → root
    }
}

File: rust/src/aho_corasick.rs
use std::collections::{HashMap, VecDeque};

const ASCII_SIZE: usize = 128;
const NO_STATE: u32 = u32::MAX;

/// Build-phase node for the Aho-Corasick automaton.
struct BuildNode<V: Clone> {
    ascii: [u32; ASCII_SIZE],
    extended: Option<HashMap<char, u32>>,
    output: Vec<V>,
}

/// A generic Aho-Corasick automaton for multi-pattern substring matching.
///
/// Uses a DFA with array-indexed transitions for ASCII characters and a
/// HashMap fallback for non-ASCII. After `build()`, the goto function is
/// fully completed so search requires no failure-link chasing.
pub struct AhoCorasick<V: Clone> {
    // Build phase
    build_nodes: Option<Vec<BuildNode<V>>>,
    empty_pattern_values: Vec<V>,
    has_patterns: bool,

    // Search phase (populated by build)
    goto_table: Vec<[u32; ASCII_SIZE]>,
    extended_goto: Vec<Option<HashMap<char, u32>>>,
    output: Vec<Box<[V]>>,
    built: bool,
}

impl<V: Clone> AhoCorasick<V> {
    /// Creates a new empty automaton.
    pub fn new() -> Self {
        let root = BuildNode {
            ascii: [NO_STATE; ASCII_SIZE],
            extended: None,
            output: Vec::new(),
        };
        Self {
            build_nodes: Some(vec![root]),
            empty_pattern_values: Vec::new(),
            has_patterns: false,
            goto_table: Vec::new(),
            extended_goto: Vec::new(),
            output: Vec::new(),
            built: false,
        }
    }

    /// Returns `true` if no patterns have been inserted.
    pub fn is_empty(&self) -> bool {
        !self.has_patterns && self.empty_pattern_values.is_empty()
    }

    /// Inserts a pattern with an associated value.
    ///
    /// # Panics
    /// Panics if called after `build()`.
    pub fn insert(&mut self, pattern: &str, value: V) {
        assert!(!self.built, "Cannot insert after build()");
        self.has_patterns = true;

        if pattern.is_empty() {
            self.empty_pattern_values.push(value);
            return;
        }

        let nodes = self.build_nodes.as_mut().unwrap();
        let mut state = 0u32;
        for c in pattern.chars() {
            let next = Self::get_goto_build(nodes, state, c);
            if next == NO_STATE {
                let new_id = nodes.len() as u32;
                Self::set_goto_build(nodes, state, c, new_id);
                nodes.push(BuildNode {
                    ascii: [NO_STATE; ASCII_SIZE],
                    extended: None,
                    output: Vec::new(),
                });
                state = new_id;
            } else {
                state = next;
            }
        }
        nodes[state as usize].output.push(value);
    }

    /// Constructs the automaton by computing failure links and completing the DFA.
    pub fn build(&mut self) {
        let nodes = self.build_nodes.take().unwrap();
        let state_count = nodes.len();

        // Copy to mutable search-phase structures
        let mut goto: Vec<[u32; ASCII_SIZE]> = nodes.iter().map(|n| n.ascii).collect();
        let mut extended: Vec<Option<HashMap<char, u32>>> =
            nodes.iter().map(|n| n.extended.clone()).collect();
        let mut output: Vec<Vec<V>> = nodes.into_iter().map(|n| n.output).collect();

        let mut failure = vec![0u32; state_count];
        let mut queue = VecDeque::new();

        // Phase 1: init depth-1 states
        for c in 0..ASCII_SIZE {
            let child = goto[0][c];
            if child == NO_STATE {
                goto[0][c] = 0; // self-loop on root
            } else {
                failure[child as usize] = 0;
                queue.push_back(child);
            }
        }
        if let Some(ref ext) = extended[0] {
            for &child in ext.values() {
                failure[child as usize] = 0;
                queue.push_back(child);
            }
        }

        // Phase 2: compute failure links
        while let Some(current) = queue.pop_front() {
            let cur = current as usize;

            for c in 0..ASCII_SIZE {
                let child = goto[cur][c];
                if child != NO_STATE {
                    let f = Self::follow_failure(&goto, &extended, &failure, current, c as u8 as char);
                    failure[child as usize] = f;
                    Self::merge_output(&mut output, child as usize, f as usize);
                    queue.push_back(child);
                }
            }

            if let Some(ext) = extended[cur].clone() {
                for (&c, &child) in &ext {
                    let f = Self::follow_failure(&goto, &extended, &failure, current, c);
                    failure[child as usize] = f;
                    Self::merge_output(&mut output, child as usize, f as usize);
                    queue.push_back(child);
                }
            }
        }

        // Phase 3: complete DFA
        // Seed with root's children
        for c in 0..ASCII_SIZE {
            let child = goto[0][c];
            if child != 0 {
                queue.push_back(child);
            }
        }
        if let Some(ref ext) = extended[0] {
            for &child in ext.values() {
                queue.push_back(child);
            }
        }

        while let Some(current) = queue.pop_front() {
            let cur = current as usize;
            let fail = failure[cur] as usize;

            for c in 0..ASCII_SIZE {
                if goto[cur][c] == NO_STATE {
                    goto[cur][c] = goto[fail][c]; // inherit from failure
                } else {
                    queue.push_back(goto[cur][c]);
                }
            }

            // Enqueue extended children BEFORE inheriting
            if let Some(ref ext) = extended[cur].clone() {
                for &child in ext.values() {
                    if child != 0 {
                        queue.push_back(child);
                    }
                }
            }

            // Inherit extended transitions from failure state
            if let Some(fail_ext) = extended[fail].clone() {
                let ext = extended[cur].get_or_insert_with(|| HashMap::with_capacity(4));
                for (c, target) in fail_ext {
                    ext.entry(c).or_insert(target);
                }
            }
        }

        self.goto_table = goto;
        self.extended_goto = extended;
        self.output = output.into_iter().map(|v| v.into_boxed_slice()).collect();
        self.built = true;
    }

    /// Searches the text and invokes the callback for each matching value.
    ///
    /// # Panics
    /// Panics (in debug builds) if `build()` has not been called.
    pub fn search(&self, text: &str, callback: &mut impl FnMut(&V)) {
        debug_assert!(self.built, "Must call build() before search()");

        for v in &self.empty_pattern_values {
            callback(v);
        }
        let mut state = 0u32;
        for c in text.chars() {
            state = self.next_state(state, c);
            for v in &*self.output[state as usize] {
                callback(v);
            }
        }
    }

    /// Byte-oriented search. Iterates `text.as_bytes()` directly, using
    /// the goto table for bytes < 128 and resetting to state 0 for
    /// bytes >= 128 (safe since all patterns are ASCII).
    pub fn search_bytes(&self, text: &str, callback: &mut impl FnMut(&V)) {
        debug_assert!(self.built, "Must call build() before search_bytes()");

        for v in &self.empty_pattern_values {
            callback(v);
        }
        let mut state = 0u32;
        for &b in text.as_bytes() {
            if b < 128 {
                state = self.goto_table[state as usize][b as usize];
            } else {
                state = 0;
            }
            for v in &*self.output[state as usize] {
                callback(v);
            }
        }
    }

    /// Searches the text and returns all matching values.
    pub fn search_collect(&self, text: &str) -> Vec<V> {
        let mut result = Vec::new();
        self.search(text, &mut |v| result.push(v.clone()));
        result
    }

    fn get_goto_build(nodes: &[BuildNode<V>], state: u32, c: char) -> u32 {
        if (c as u32) < ASCII_SIZE as u32 {
            nodes[state as usize].ascii[c as usize]
        } else {
            nodes[state as usize]
                .extended
                .as_ref()
                .and_then(|m| m.get(&c).copied())
                .unwrap_or(NO_STATE)
        }
    }

    fn set_goto_build(nodes: &mut [BuildNode<V>], state: u32, c: char, target: u32) {
        if (c as u32) < ASCII_SIZE as u32 {
            nodes[state as usize].ascii[c as usize] = target;
        } else {
            nodes[state as usize]
                .extended
                .get_or_insert_with(|| HashMap::with_capacity(4))
                .insert(c, target);
        }
    }

    fn follow_failure(
        goto: &[[u32; ASCII_SIZE]],
        extended: &[Option<HashMap<char, u32>>],
        failure: &[u32],
        parent: u32,
        c: char,
    ) -> u32 {
        let mut state = failure[parent as usize];
        while state != 0 && Self::get_goto_search(goto, extended, state, c) == NO_STATE {
            state = failure[state as usize];
        }
        let target = Self::get_goto_search(goto, extended, state, c);
        if target != NO_STATE { target } else { 0 }
    }

    fn get_goto_search(
        goto: &[[u32; ASCII_SIZE]],
        extended: &[Option<HashMap<char, u32>>],
        state: u32,
        c: char,
    ) -> u32 {
        if (c as u32) < ASCII_SIZE as u32 {
            goto[state as usize][c as usize]
        } else {
            extended[state as usize]
                .as_ref()
                .and_then(|m| m.get(&c).copied())
                .unwrap_or(NO_STATE)
        }
    }

    fn merge_output(output: &mut [Vec<V>], state: usize, fail_state: usize) {
        if output[fail_state].is_empty() {
            return;
        }
        let fail_out = output[fail_state].clone();
        output[state].extend(fail_out);
    }

    fn next_state(&self, state: u32, c: char) -> u32 {
        if (c as u32) < ASCII_SIZE as u32 {
            self.goto_table[state as usize][c as usize]
        } else {
            self.extended_goto[state as usize]
                .as_ref()
                .and_then(|m| m.get(&c).copied())
                .unwrap_or(0) // unknown non-ASCII → root
        }
    }
}

impl<V: Clone> Default for AhoCorasick<V> {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // --- Generic AhoCorasick<String> tests (from AhoCorasickTest.java) ---

    #[test]
    fn finds_single_pattern() {
        let mut ac = AhoCorasick::new();
        ac.insert("he", "val".to_string());
        ac.build();
        let result = ac.search_collect("she");
        assert!(result.contains(&"val".to_string()));
    }

    #[test]
    fn finds_multiple_patterns() {
        let mut ac = AhoCorasick::new();
        ac.insert("he", "v1".to_string());
        ac.insert("she", "v2".to_string());
        ac.insert("his", "v3".to_string());
        ac.insert("hers", "v4".to_string());
        ac.build();

        let result = ac.search_collect("shers");
        assert!(result.contains(&"v1".to_string()), "should find 'he'");
        assert!(result.contains(&"v2".to_string()), "should find 'she'");
        assert!(result.contains(&"v4".to_string()), "should find 'hers'");
        assert!(!result.contains(&"v3".to_string()), "should not find 'his'");
    }

    #[test]
    fn finds_overlapping_patterns() {
        let mut ac = AhoCorasick::new();
        ac.insert("ab", "v1".to_string());
        ac.insert("bc", "v2".to_string());
        ac.build();
        let result = ac.search_collect("abc");
        assert!(result.contains(&"v1".to_string()));
        assert!(result.contains(&"v2".to_string()));
    }

    #[test]
    fn no_match_returns_empty() {
        let mut ac = AhoCorasick::new();
        ac.insert("xyz", "val".to_string());
        ac.build();
        let result = ac.search_collect("abc");
        assert!(result.is_empty());
    }

    #[test]
    #[should_panic(expected = "Must call build()")]
    fn panics_if_search_before_build() {
        let mut ac = AhoCorasick::new();
        ac.insert("test", "val".to_string());
        ac.search_collect("test");
    }

    #[test]
    #[should_panic(expected = "Cannot insert after build()")]
    fn panics_if_insert_after_build() {
        let mut ac: AhoCorasick<String> = AhoCorasick::new();
        ac.build();
        ac.insert("test", "val".to_string());
    }

    #[test]
    fn empty_pattern_matches_any_text() {
        let mut ac = AhoCorasick::new();
        ac.insert("", "empty".to_string());
        ac.build();
        let result = ac.search_collect("anything");
        assert!(result.contains(&"empty".to_string()));
    }

    #[test]
    fn finds_pattern_at_end() {
        let mut ac = AhoCorasick::new();
        ac.insert("sport", "val".to_string());
        ac.build();
        let result = ac.search_collect("/category/sport");
        assert!(result.contains(&"val".to_string()));
    }

    #[test]
    fn finds_pattern_in_middle() {
        let mut ac = AhoCorasick::new();
        ac.insert("sport", "val".to_string());
        ac.build();
        let result = ac.search_collect("/category/sport/items");
        assert!(result.contains(&"val".to_string()));
    }

    #[test]
    fn non_ascii_pattern() {
        let mut ac = AhoCorasick::new();
        ac.insert("\u{00E9}l\u{00E8}ve", "found".to_string());
        ac.build();
        let result = ac.search_collect("un \u{00E9}l\u{00E8}ve ici");
        assert!(result.contains(&"found".to_string()));
    }

    // --- AhoCorasick<u32> tests (from IntAhoCorasickTest.java) ---

    fn search_u32(ac: &AhoCorasick<u32>, text: &str) -> Vec<u32> {
        ac.search_collect(text)
    }

    #[test]
    fn int_finds_single_pattern() {
        let mut ac = AhoCorasick::new();
        ac.insert("he", 1u32);
        ac.build();
        assert!(search_u32(&ac, "she").contains(&1));
    }

    #[test]
    fn int_finds_multiple_patterns() {
        let mut ac = AhoCorasick::new();
        ac.insert("he", 1u32);
        ac.insert("she", 2u32);
        ac.insert("his", 3u32);
        ac.insert("hers", 4u32);
        ac.build();

        let result = search_u32(&ac, "shers");
        assert!(result.contains(&1), "should find 'he'");
        assert!(result.contains(&2), "should find 'she'");
        assert!(result.contains(&4), "should find 'hers'");
        assert!(!result.contains(&3), "should not find 'his'");
    }

    #[test]
    fn int_finds_overlapping_patterns() {
        let mut ac = AhoCorasick::new();
        ac.insert("ab", 1u32);
        ac.insert("bc", 2u32);
        ac.build();
        let result = search_u32(&ac, "abc");
        assert!(result.contains(&1));
        assert!(result.contains(&2));
    }

    #[test]
    fn int_no_match_returns_empty() {
        let mut ac = AhoCorasick::new();
        ac.insert("xyz", 1u32);
        ac.build();
        assert!(search_u32(&ac, "abc").is_empty());
    }

    #[test]
    #[should_panic(expected = "Must call build()")]
    fn int_panics_if_search_before_build() {
        let mut ac = AhoCorasick::new();
        ac.insert("test", 1u32);
        ac.search("test", &mut |_| {});
    }

    #[test]
    #[should_panic(expected = "Cannot insert after build()")]
    fn int_panics_if_insert_after_build() {
        let mut ac: AhoCorasick<u32> = AhoCorasick::new();
        ac.build();
        ac.insert("test", 1u32);
    }

    #[test]
    fn int_empty_pattern_matches_any_text() {
        let mut ac = AhoCorasick::new();
        ac.insert("", 42u32);
        ac.build();
        assert!(search_u32(&ac, "anything").contains(&42));
    }

    #[test]
    fn int_finds_pattern_at_end() {
        let mut ac = AhoCorasick::new();
        ac.insert("sport", 1u32);
        ac.build();
        assert!(search_u32(&ac, "/category/sport").contains(&1));
    }

    #[test]
    fn int_finds_pattern_in_middle() {
        let mut ac = AhoCorasick::new();
        ac.insert("sport", 1u32);
        ac.build();
        assert!(search_u32(&ac, "/category/sport/items").contains(&1));
    }

    #[test]
    fn int_is_empty_when_new() {
        assert!(AhoCorasick::<u32>::new().is_empty());
    }

    #[test]
    fn int_is_not_empty_after_insert() {
        let mut ac = AhoCorasick::new();
        ac.insert("test", 1u32);
        assert!(!ac.is_empty());
    }

    #[test]
    fn int_is_not_empty_after_empty_pattern_insert() {
        let mut ac = AhoCorasick::new();
        ac.insert("", 1u32);
        assert!(!ac.is_empty());
    }

    #[test]
    fn int_non_ascii_pattern() {
        let mut ac = AhoCorasick::new();
        ac.insert("\u{00E9}l\u{00E8}ve", 1u32);
        ac.build();
        assert!(search_u32(&ac, "un \u{00E9}l\u{00E8}ve ici").contains(&1));
    }

    #[test]
    fn int_multiple_empty_pattern_values() {
        let mut ac = AhoCorasick::new();
        ac.insert("", 1u32);
        ac.insert("", 2u32);
        ac.insert("", 3u32);
        ac.build();
        let result = search_u32(&ac, "text");
        assert_eq!(3, result.len());
        assert!(result.contains(&1));
        assert!(result.contains(&2));
        assert!(result.contains(&3));
    }

    #[test]
    fn int_failure_link_merges_output() {
        let mut ac = AhoCorasick::new();
        ac.insert("abc", 1u32);
        ac.insert("bc", 2u32);
        ac.insert("c", 3u32);
        ac.build();

        let result = search_u32(&ac, "abc");
        assert!(result.contains(&1));
        assert!(result.contains(&2));
        assert!(result.contains(&3));
    }

    #[test]
    fn int_many_patterns_stress_test() {
        let mut ac = AhoCorasick::new();
        for i in 0..100u32 {
            ac.insert(&format!("pattern{}", i), i);
        }
        ac.build();
        let result = search_u32(&ac, "this has pattern42 and pattern7 inside");
        assert!(result.contains(&42));
        assert!(result.contains(&7));
    }

    #[test]
    fn search_bytes_finds_single_pattern() {
        let mut ac = AhoCorasick::new();
        ac.insert("he", 1u32);
        ac.build();
        let mut result = Vec::new();
        ac.search_bytes("she", &mut |v| result.push(*v));
        assert!(result.contains(&1));
    }

    #[test]
    fn search_bytes_finds_multiple_patterns() {
        let mut ac = AhoCorasick::new();
        ac.insert("he", 1u32);
        ac.insert("she", 2u32);
        ac.insert("his", 3u32);
        ac.insert("hers", 4u32);
        ac.build();

        let mut result = Vec::new();
        ac.search_bytes("shers", &mut |v| result.push(*v));
        assert!(result.contains(&1), "should find 'he'");
        assert!(result.contains(&2), "should find 'she'");
        assert!(result.contains(&4), "should find 'hers'");
        assert!(!result.contains(&3), "should not find 'his'");
    }

    #[test]
    fn search_bytes_empty_pattern() {
        let mut ac = AhoCorasick::new();
        ac.insert("", 42u32);
        ac.build();
        let mut result = Vec::new();
        ac.search_bytes("anything", &mut |v| result.push(*v));
        assert!(result.contains(&42));
    }
}
